zulipterminal/helper.py

```diff
@@ -459,15 +460,71 @@ def match_emoji(emoji: str, text: str) -> bool:
     return emoji.lower().startswith(text.lower())
 
 
-def match_stream(stream: Any, text: str) -> bool:
+StreamT = TypeVar('StreamT')
+
+
+def match_stream(streams: List[Tuple[StreamT, str]], text: str,
+                 pinned_streams: List[List[Any]]) -> List[StreamT]:
     """
-    True if the stream matches with `text` (case insensitive),
-    False otherwise.
+    Returns a list of streams (StreamT) whose words match with the 'text' in
+    the following order:
+    * 1st-word startswith match > 2nd-word startswith match > ... (pinned)
+    * 1st-word startswith match > 2nd-word startswith match > ... (unpinned)
+
+    Data structures:
+    * `stream_splits` is a list of tuples where each tuple T ->
+      (stream, [stream_name, second_word, third_word, ...]).
+    * `matches` is an OrderedDict which stores streams (StreamT) that match in
+      the following manner:
+      matches = {
+          'pinned': {
+              0: [streams (StreamT) whose first word matches],  # If any.
+              1: [streams (StreamT) whose second word matches], # If any.
+              ...
+          },
+          # Similarly for 'unpinned'.
+      }
+
+    Note: This function expects `streams` to be sorted in a non-decreasing
+    order.
     """
-    stream_name = stream[0].lower()
-    if stream_name.startswith(text.lower()):
-        return True
-    return False
+    pinned_stream_names = [stream[0] for stream in pinned_streams]
+
+    # Note that 'space' is put in `delimiters` only to show the types of
+    # delimiters that are supported. Otherwise, it is redundant (see the
+    # following .replace method).
+    delimiters = [' ', '-', '_', '/']
+    stream_splits = []
+    for stream, stream_name in streams:
+        delimit_with_spaces = stream_name
+        for delimiter in delimiters:
+            delimit_with_spaces = delimit_with_spaces.replace(delimiter, ' ')
```

Comment:
This might be clearer with maketrans or a regex? You might then be able to shrink this into a list comprehension?

Reply:
Wow, I didn't know something like `maketrans` existed. Thanks!